File: sdks/typescript/src/utils/http.ts
// HTTP Client Wrapper for OpenAgentMail

import type { OpenAgentMailConfig, RequestOptions, ApiErrorResponse } from '../types.js';
import {
  APIError,
  BadRequestError,
  AuthenticationError,
  AuthorizationError,
  NotFoundError,
  ConflictError,
  ValidationError,
  RateLimitError,
  InternalServerError,
  TimeoutError,
  NetworkError,
} from '../errors.js';

const DEFAULT_BASE_URL = 'https://api.openagentmail.com/v0';
const DEFAULT_TIMEOUT = 30000;

/**
 * Convert camelCase to snake_case for API requests
 */
export function toSnakeCase(str: string): string {
  return str.replace(/[A-Z]/g, (letter) => `_${letter.toLowerCase()}`);
}

/**
 * Convert snake_case to camelCase for API responses
 */
export function toCamelCase(str: string): string {
  return str.replace(/_([a-z])/g, (_, letter) => letter.toUpperCase());
}

/**
 * Recursively transform object keys from camelCase to snake_case
 */
export function transformRequestBody(obj: unknown): unknown {
  if (obj === null || obj === undefined) {
    return obj;
  }
  
  if (Array.isArray(obj)) {
    return obj.map(transformRequestBody);
  }
  
  if (typeof obj === 'object') {
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(obj as Record<string, unknown>)) {
      result[toSnakeCase(key)] = transformRequestBody(value);
    }
    return result;
  }
  
  return obj;
}

/**
 * Recursively transform object keys from snake_case to camelCase
 */
export function transformResponseBody<T>(obj: unknown): T {
  if (obj === null || obj === undefined) {
    return obj as T;
  }
  
  if (Array.isArray(obj)) {
    return obj.map(transformResponseBody) as T;
  }
  
  if (typeof obj === 'object') {
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(obj as Record<string, unknown>)) {
      result[toCamelCase(key)] = transformResponseBody(value);
    }
    return result as T;
  }
  
  return obj as T;
}

/**
 * Build query string from parameters
 */
function buildQueryString(params: Record<string, string | number | boolean | undefined>): string {
  const searchParams = new URLSearchParams();
  
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined) {
      searchParams.set(toSnakeCase(key), String(value));
    }
  }
  
  const queryString = searchParams.toString();
  return queryString ? `?${queryString}` : '';
}

/**
 * HTTP client for making API requests
 */
export class HttpClient {
  private readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly timeout: number;
  private readonly fetchFn: typeof fetch;

  constructor(config: OpenAgentMailConfig) {
    this.apiKey = config.apiKey;
    this.baseUrl = config.baseUrl ?? DEFAULT_BASE_URL;
    this.timeout = config.timeout ?? DEFAULT_TIMEOUT;
    this.fetchFn = config.fetch ?? fetch;
  }

  /**
   * Make an HTTP request to the API
   */
  async request<T>(options: RequestOptions): Promise<T> {
    const { method, path, body, query } = options;
    
    let url = `${this.baseUrl}${path}`;
    if (query) {
      url += buildQueryString(query);
    }

    const headers: Record<string, string> = {
      'Authorization': `Bearer ${this.apiKey}`,
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };

    const requestInit: RequestInit = {
      method,
      headers,
    };

    if (body !== undefined) {
      requestInit.body = JSON.stringify(transformRequestBody(body));
    }

    // Create abort controller for timeout
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);
    requestInit.signal = controller.signal;

    let response: Response;
    
    try {
      response = await this.fetchFn(url, requestInit);
    } catch (error) {
      clearTimeout(timeoutId);
      
      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          throw new TimeoutError(this.timeout);
        }
        throw new NetworkError(`Network request failed: ${error.message}`, error);
      }
      
      throw new NetworkError('Network request failed');
    } finally {
      clearTimeout(timeoutId);
    }

    // Handle 204 No Content
    if (response.status === 204) {
      return undefined as T;
    }

    let responseBody: unknown;
    
    try {
      responseBody = await response.json();
    } catch {
      if (!response.ok) {
        throw new APIError(response.status, {
          code: 'unknown_error',
          message: `Request failed with status ${response.status}`,
        });
      }
      return undefined as T;
    }

    if (!response.ok) {
      const errorResponse = responseBody as ApiErrorResponse;
      const error = errorResponse.error ?? {
        code: 'unknown_error',
        message: `Request failed with status ${response.status}`,
      };

      switch (response.status) {
        case 400:
          throw new BadRequestError(error);
        case 401:
          throw new AuthenticationError(error);
        case 403:
          throw new AuthorizationError(error);
        case 404:
          throw new NotFoundError(error);
        case 409:
          throw new ConflictError(error);
        case 422:
          throw new ValidationError(error);
        case 429:
          throw new RateLimitError(error, response.headers);
        default:
          if (response.status >= 500) {
            throw new InternalServerError(response.status, error);
          }
          throw new APIError(response.status, error);
      }
    }

    return transformResponseBody<T>(responseBody);
  }

  /**
   * Make a GET request
   */
  get<T>(path: string, query?: Record<string, string | number | boolean | undefined>): Promise<T> {
    return this.request<T>({ method: 'GET', path, query });
  }

  /**
   * Make a POST request
   */
  post<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>({ method: 'POST', path, body });
  }

  /**
   * Make a PUT request
   */
  put<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>({ method: 'PUT', path, body });
  }

  /**
   * Make a PATCH request
   */
  patch<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>({ method: 'PATCH', path, body });
  }

  /**
   * Make a DELETE request
   */
  delete<T>(path: string): Promise<T> {
    return this.request<T>({ method: 'DELETE', path });
  }
}

File: sdks/python/openagentmail/_utils/http.py
"""HTTP client utilities for OpenAgentMail SDK."""

from __future__ import annotations

from typing import Any, Optional, Type, TypeVar

import httpx

from ..errors import (
    APIError,
    AuthenticationError,
    AuthorizationError,
    ConflictError,
    ConnectionError,
    InternalServerError,
    NotFoundError,
    RateLimitError,
    TimeoutError,
    ValidationError,
)

T = TypeVar("T")

DEFAULT_BASE_URL = "https://api.openagentmail.com/v0"
DEFAULT_TIMEOUT = 30.0


class HTTPClient:
    """Synchronous HTTP client for API requests."""

    def __init__(
        self,
        api_key: str,
        base_url: str = DEFAULT_BASE_URL,
        timeout: float = DEFAULT_TIMEOUT,
    ) -> None:
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self._client: Optional[httpx.Client] = None

    @property
    def client(self) -> httpx.Client:
        if self._client is None:
            self._client = httpx.Client(
                base_url=self.base_url,
                headers={
                    "Authorization": f"Bearer {self.api_key}",
                    "Content-Type": "application/json",
                    "User-Agent": "openagentmail-python/0.1.0",
                },
                timeout=self.timeout,
            )
        return self._client

    def close(self) -> None:
        if self._client is not None:
            self._client.close()
            self._client = None

    def __enter__(self) -> "HTTPClient":
        return self

    def __exit__(self, *args: Any) -> None:
        self.close()

    def request(
        self,
        method: str,
        path: str,
        json: Optional[dict[str, Any]] = None,
        params: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        """Make an HTTP request and return JSON response."""
        # Filter out None values from params
        if params:
            params = {k: v for k, v in params.items() if v is not None}

        try:
            response = self.client.request(
                method=method,
                url=path,
                json=json,
                params=params,
            )
        except httpx.ConnectError as e:
            raise ConnectionError(f"Failed to connect: {e}") from e
        except httpx.TimeoutException as e:
            raise TimeoutError(f"Request timed out: {e}") from e

        return self._handle_response(response)

    def _handle_response(self, response: httpx.Response) -> dict[str, Any]:
        """Handle response and raise appropriate errors."""
        if response.status_code == 204:
            return {}

        try:
            data = response.json()
        except Exception:
            if response.is_success:
                return {}
            raise InternalServerError(f"Invalid JSON response: {response.text}")

        if response.is_success:
            return data

        # Extract error info
        error_data = data.get("error", {})
        message = error_data.get("message", "Unknown error")
        code = error_data.get("code", "unknown_error")
        details = error_data.get("details")

        # Map to specific exception types
        status = response.status_code
        if status == 401:
            raise AuthenticationError(message)
        elif status == 403:
            raise AuthorizationError(message)
        elif status == 404:
            raise NotFoundError(message)
        elif status == 409:
            raise ConflictError(message)
        elif status == 422:
            raise ValidationError(message, details)
        elif status == 429:
            retry_after = response.headers.get("Retry-After")
            raise RateLimitError(
                message,
                retry_after=int(retry_after) if retry_after else None,
            )
        elif status >= 500:
            raise InternalServerError(message)
        else:
            raise APIError(message, status, code, details)

    def get(
        self,
        path: str,
        params: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        return self.request("GET", path, params=params)

    def post(
        self,
        path: str,
        json: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        return self.request("POST", path, json=json)

    def put(
        self,
        path: str,
        json: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        return self.request("PUT", path, json=json)

    def patch(
        self,
        path: str,
        json: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        return self.request("PATCH", path, json=json)

    def delete(
        self,
        path: str,
    ) -> dict[str, Any]:
        return self.request("DELETE", path)


class AsyncHTTPClient:
    """Asynchronous HTTP client for API requests."""

    def __init__(
        self,
        api_key: str,
        base_url: str = DEFAULT_BASE_URL,
        timeout: float = DEFAULT_TIMEOUT,
    ) -> None:
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self._client: Optional[httpx.AsyncClient] = None

    @property
    def client(self) -> httpx.AsyncClient:
        if self._client is None:
            self._client = httpx.AsyncClient(
                base_url=self.base_url,
                headers={
                    "Authorization": f"Bearer {self.api_key}",
                    "Content-Type": "application/json",
                    "User-Agent": "openagentmail-python/0.1.0",
                },
                timeout=self.timeout,
            )
        return self._client

    async def close(self) -> None:
        if self._client is not None:
            await self._client.aclose()
            self._client = None

    async def __aenter__(self) -> "AsyncHTTPClient":
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self.close()

    async def request(
        self,
        method: str,
        path: str,
        json: Optional[dict[str, Any]] = None,
        params: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        """Make an async HTTP request and return JSON response."""
        # Filter out None values from params
        if params:
            params = {k: v for k, v in params.items() if v is not None}

        try:
            response = await self.client.request(
                method=method,
                url=path,
                json=json,
                params=params,
            )
        except httpx.ConnectError as e:
            raise ConnectionError(f"Failed to connect: {e}") from e
        except httpx.TimeoutException as e:
            raise TimeoutError(f"Request timed out: {e}") from e

        return self._handle_response(response)

    def _handle_response(self, response: httpx.Response) -> dict[str, Any]:
        """Handle response and raise appropriate errors."""
        if response.status_code == 204:
            return {}

        try:
            data = response.json()
        except Exception:
            if response.is_success:
                return {}
            raise InternalServerError(f"Invalid JSON response: {response.text}")

        if response.is_success:
            return data

        # Extract error info
        error_data = data.get("error", {})
        message = error_data.get("message", "Unknown error")
        code = error_data.get("code", "unknown_error")
        details = error_data.get("details")

        # Map to specific exception types
        status = response.status_code
        if status == 401:
            raise AuthenticationError(message)
        elif status == 403:
            raise AuthorizationError(message)
        elif status == 404:
            raise NotFoundError(message)
        elif status == 409:
            raise ConflictError(message)
        elif status == 422:
            raise ValidationError(message, details)
        elif status == 429:
            retry_after = response.headers.get("Retry-After")
            raise RateLimitError(
                message,
                retry_after=int(retry_after) if retry_after else None,
            )
        elif status >= 500:
            raise InternalServerError(message)
        else:
            raise APIError(message, status, code, details)

    async def get(
        self,
        path: str,
        params: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        return await self.request("GET", path, params=params)

    async def post(
        self,
        path: str,
        json: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        return await self.request("POST", path, json=json)

    async def put(
        self,
        path: str,
        json: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        return await self.request("PUT", path, json=json)

    async def patch(
        self,
        path: str,
        json: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        return await self.request("PATCH", path, json=json)

    async def delete(
        self,
        path: str,
    ) -> dict[str, Any]:
        return await self.request("DELETE", path)
